dds_web/database/models.py

```diff
@@ -491,6 +499,92 @@ def verify_HOTP(self, token):
         self.hotp_issue_time = None
         db.session.commit()
 
+    @property
+    def totp_initiated(self):
+        """To check if activation of TOTP has been initiated, not to be confused
+        with user.totp_enabled, that indicates if TOTP is successfully enabled for the user."""
+        return self._totp_secret is not None
+
+    def totp_object(self):
+        """Google Authenticator seems to only be able to handle SHA1 and 6 digit codes"""
+        if self.totp_initiated:
+            return twofactor_totp.TOTP(self._totp_secret, 6, hashes.SHA1(), 30)
+        return None
+
+    def setup_totp_secret(self):
+        """Generate random 160 bit as the new totp secret and return provisioning URI
+        We're using SHA1 (Google Authenticator seems to only use SHA1 and 6 digit codes)
+        so secret should be at least 160 bits
+        https://cryptography.io/en/latest/hazmat/primitives/twofactor/#cryptography.hazmat.primitives.twofactor.totp.TOTP
+        """
+        self._totp_secret = os.urandom(20)
+        db.session.commit()
+
+    @property
+    def totp_secret_and_uri(self):
+        """Returns the users totp provisioning URI. Can only be sent before totp has been enabled."""
+        if self.totp_enabled:
+            # Can not be fetched again after it has been enabled
+            raise AuthenticationError("TOTP secret already enabled.")
+        totp = self.totp_object()
+
+        return self._totp_secret, totp.get_provisioning_uri(
+            account_name=self.username,
+            issuer="Data Delivery System",
+        )
+
+    def activate_totp(self):
+        """Set TOTP as the preferred means of second factor authentication.
+        Should be called after first totp token is verified
+        """
+        self.totp_enabled = True
+        db.session.commit()
+
+    def deactivate_totp(self):
+        """Fallback to HOTP as the preferred means of second factor authentication."""
+        self.totp_enabled = False
+        db.session.commit()
+
+    def verify_TOTP(self, token):
+        """Verify the totp token. Checks the previous, current and comming time frame
+        to allow for some clock drift.
+
+        raises AuthenticationError if token is invalid, has expired or
+        if totp has been successfully verified within the last 90 seconds.
+        """
+        # can't use totp successfully more than once within 90 seconds.
+        # Time frame chosen so that no one can use the same token more than once
+        # No need to use epoch time here.
+        current_time = dds_web.utils.current_time()
+        if self.totp_last_verified and (
```

Comment:
Should it be `and`? Not sure I follow?

Reply:
Yeah, I think it's a check to make sure `totp_last_verified` is not null, that is, if totp hasn't been used yet. So the check is: "have TOTP been used before and was it too recently used"